README.md
# SartoriUSB

[![Build Status](https://travis-ci.com/holgi/sartoriusb.svg?branch=master)](https://travis-ci.com/holgi/sartoriusb)

USB serial connection to Sartorius Quintix Scales

## Example:

```python

    import sartoriusb as sub

    with sub.SartoriusUsb('COM4') as conn:
        result = conn.measure()

    assert result.mode == "G"
    assert result.value == "+3.4"
    assert result.unit == "mg"
    assert result.stable == True
    assert result.message == None
```

## Required settings on the scale

The USB protocol on the scale must be set to SBI.

## Connection to the scale

The prefered way to establish an connection is the use of a context manager:
`with sartoriusb.SartoriusUsb('COM4') as conn:`

For manually handling the connection, the methods `connect()` and
`close()` can be used. For convenience there is also and `open()`
method that just behaves like `connect()`


## Sending commands and receiving output

Sending commands to the scale can be done with the `send(cmd)`, the
available commands according to the manual are listed below, but you could
but propably shouldn't send any arbitrary commands to the scale.

For reading the results from the scale, there are two methods available:
- `read(bytes)`: reads a number of bytes from the scale
- `readlines()`: returns a list of text lines, already decoded from bytes

Since sending a command and recieving an output are often used side by side,
there is a convenience method: `get(cmd)` will first send the command and
returns the output from the scale as a list of strings.

## Measuring

To measure a weight (or whatever the scale is set to), you could use
`get(CMD_PRINT)` and parse the output yourselve or use the `measure()`
method and get an already parsed output:

The function returns a `sartoriusb.Measurement` named tuple:

- `Measurement.mode`:
   The measurement mode, might be `unknown` if reporting on the scale is set
   to 16 byte output
- `Measurement.value`:
    The value as string as reported by the scale; whitespace an nonnumeric
    characters are removed, so `int()` or `float()` could be used directly
- `Measurement.unit`:
    The unit of the Measurement reported by the scale. Might be `None` if it
    was an unstable read
- `Measurement.stable`:
    Boolean value indicating if the measured value was stable or unstable
- `Measurement.message`:
    If a non-measurement message was sent from the scale, it is reported here
    and all other fields are set to `None`. For a regular measurement result
    this field is `None`.

## Available Commands

- `CMD_PRINT`: "Print" the current measurement result
- `CMD_TARA`: performa a tara or set the display to zero

- `CMD_EXPCLICIT_TARA`: explicit tara only
- `CMD_EXPCLICIT_NULL`: explicit set display to zero

- `CMD_INFO_TYPE`:  show scale type
- `CMD_INFO_SNR`: show serial number of scale
- `CMD_INFO_VERSION_SCALE`: show software version of the scale
- `CMD_INFO_VERSION_CONTROL_UNIT`: show the software version of the command and display module
- `CMD_INFO_USER`: display user and device id

- `CMD_FILTER_ENVIRONMENT_VERY_STABLE`: set the filter to "very stable enviroment"
- `CMD_FILTER_ENVIRONMENT_STABLE`: set the filter to "stable enviroment"
- `CMD_FILTER_ENVIRONMENT_UNSTABLE`:  set the filter to "unstable enviroment"
- `CMD_FILTER_ENVIRONMENT_VERAY_UNSTABLE`:  set the filter to "very unstable enviroment"

- `CMD_KEYBOARD_LOCK`: lock the keyboard on the command and display module
- `CMD_KEYBOARD_UNKLOCK`: unlock the keyboard on the command and display module
- `CMD_KEYPRESS_PRINT`: simulate a keypress on the command modules 'print' key
- `CMD_KEYPRESS_CANCEL`: simulate a keypress on the command modules 'cancel' key

- `CMD_BEEP`: make the scale beep

- `CMD_ADJUST_INTERNAL`: start internal adjustment
- `CMD_ADJUST_EXTERNAL`: start external adjustment with standard weights

- `CMD_RESTART`: restart the scale



## Development

To install the development version of SartoriUSB:

    git clone https://github.com/holgi/sartoriusb.git
    cd sartoriusb

    # intialize the developement environment and activate it
    make devenv
    .venv/bin/activate

To run the tests, run `make tests` or `make coverage` for a complete report.


CHANGES.md
0.2.5 - heartbeat update
------------------------

  - no changes, just a new version to show I still care about it


0.2.4 - heartbeat update
------------------------

  - no changes, just a new version to show I still care about it


0.2.3 - setup tuning
--------------------

 - more commands added to makefile
 - linting improved with isort


0.2.0 - travis ci
-----------------

 - added ci integration with travis


0.1.0 - first release
---------------------

 - first somehow working library


tests/test_sartoriusb.py
from unittest.mock import call

import pytest


def test_sartorius_usb_init():
    from sartoriusb import SartoriusUsb

    sub = SartoriusUsb(1, 2, timeout=3, something=4, different=6)

    assert sub._serial_args == (1, 2)
    assert sub._serial_kargs == {"timeout": 3, "something": 4, "different": 6}
    assert sub._con is None


@pytest.mark.parametrize("con_value,expected", [(None, 1), ("Something", 0)])
def test_sartorius_usb_connection_property(mocker, con_value, expected):
    from sartoriusb import SartoriusUsb

    mocker.patch.object(SartoriusUsb, "connect")
    sub = SartoriusUsb()
    sub._con = con_value

    con = sub.connection

    assert sub.connect.call_count == expected
    assert con == con_value


def test_sartorius_connect_usb_parameters_passed_to_serial(mocker):
    from sartoriusb import SartoriusUsb, serial

    mocker.patch.object(serial, "Serial", return_value="dummy connection")
    sub = SartoriusUsb(1, 2, something=4, different=6)

    sub.connect()

    assert serial.Serial.call_count == 1
    assert serial.Serial.call_args == call(
        1, 2, timeout=1, something=4, different=6
    )
    assert sub._con == "dummy connection"


def test_sartorius_connect_on_alredy_established_connection(mocker):
    from sartoriusb import SartoriusUsb, serial

    mocker.patch.object(serial, "Serial")
    sub = SartoriusUsb()
    sub._con = "alredy connected"

    sub.connect()

    assert serial.Serial.call_count == 0


def test_sartorius_open_calls_connect(mocker):
    from sartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    sub.connect = mocker.Mock()

    sub.open()

    assert sub.connect.call_count == 1


def test_sartorius_usb_close_calls_close_on_established_connection(mocker):
    from sartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    sub._con = mocker.Mock()
    mocked_con = sub._con

    sub.close()

    assert sub._con is None
    assert mocked_con.close.call_count == 1


@pytest.mark.parametrize(
    "value,expected", [(b"bytes", b"bytes"), ("string", b"string")]
)
def test_sartorius_sends_bytes(mocker, value, expected):
    from sartoriusb import CR, LF, ESC, SartoriusUsb

    sub = SartoriusUsb()
    sub._con = mocker.Mock()

    sub.send(value)

    assert sub._con.write.call_count == 1
    assert sub._con.write.call_args == call(ESC + expected + CR + LF)


def test_sartorius_read_passes_argument(mocker):
    from sartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    sub._con = mocker.Mock()
    sub._con.read.return_value = "Hello!"

    result = sub.read(42)

    assert sub._con.read.call_count == 1
    assert sub._con.read.call_args == call(42)
    assert result == "Hello!"


def test_sartorius_readline(mocker):
    from sartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    sub._con = mocker.Mock()
    sub._con.readline.return_value = "Hello!"

    result = sub.readline()

    assert sub._con.readline.call_count == 1
    assert sub._con.readline.call_args == call()
    assert result == "Hello!"


@pytest.mark.parametrize("stop_line", [b"", b" "])
@pytest.mark.parametrize(
    "stop_position,expected",
    [(0, []), (1, ["one"]), (2, ["one", "two"]), (3, ["one", "two", "three"])],
)
def test_sartorius_readlines(mocker, stop_line, stop_position, expected):
    from sartoriusb import SartoriusUsb

    data = [b"one", b"two", b"three", b"four", b"five"]
    data[stop_position] = stop_line

    sub = SartoriusUsb()
    sub._con = mocker.Mock()
    sub._con.readline.side_effect = data

    result = sub.readlines()

    assert sub._con.readline.call_count == stop_position + 1
    assert result == expected


def test_sartorius_get(mocker):
    from sartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    mocker.patch.object(sub, "send")
    mocker.patch.object(sub, "readlines", return_value="some result")

    result = sub.get("some command")

    assert sub.send.call_count == 1
    assert sub.send.call_args == call("some command")
    assert sub.readlines.call_count == 1
    assert sub.readlines.call_args == call()
    assert result == "some result"


def test_sartorius_measure_with_data(mocker):
    from sartoriusb import CMD_PRINT, SartoriusUsb

    sub = SartoriusUsb()
    mocker.patch.object(sub, "get", return_value=["value", "some other stuff"])
    mocker.patch("sartoriusb.parse_measurement", return_value="parsed result")

    result = sub.measure()

    assert sub.get.call_count == 1
    assert sub.get.call_args == call(CMD_PRINT)
    assert result == "parsed result"


def test_sartorius_measure_with_timeout(mocker):
    from sartoriusb import CMD_PRINT, SartoriusUsb

    sub = SartoriusUsb()
    mocker.patch.object(sub, "get", return_value=[])
    mocker.patch("sartoriusb.parse_measurement", return_value="parsed result")

    result = sub.measure()

    assert sub.get.call_count == 1
    assert sub.get.call_args == call(CMD_PRINT)
    assert result == (None, None, None, None, "Connection Timeout")


def test_sartorius_context_manager(mocker):
    from sartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    mocker.patch.object(sub, "connect")
    mocker.patch.object(sub, "close")

    with sub as context:
        assert sub is context
        assert sub.connect.call_count == 1
        assert sub.close.call_count == 0

    assert sub.connect.call_count == 1
    assert sub.close.call_count == 1


@pytest.mark.parametrize(
    "value,count_16,count_22,expected",
    [
        ("123456789012345", 1, 0, "16"),
        ("1234567890123456", 1, 0, "16"),
        ("12345678901234567", 0, 1, "22"),
    ],
)
def test_sartorius_parse_measurement(
    mocker, value, count_16, count_22, expected
):
    from sartoriusb import parse_measurement

    mocker.patch("sartoriusb._parse_16_char_output", return_value="16")
    mocker.patch("sartoriusb._parse_22_char_output", return_value="22")

    result = parse_measurement(value)

    assert result == expected


def test_sartorius_parse_22_char_output(mocker):
    from sartoriusb import _parse_22_char_output

    mocker.patch("sartoriusb._parse_16_char_output", return_value="16")

    result = _parse_22_char_output(" mode value ")

    assert result == "16"


@pytest.mark.parametrize(
    "value,expected",
    [
        ("12345", False),
        (" some text ", False),
        (" Error 123 ", True),
        ("highlander ", True),
        (" too low", True),
        ("123--456", True),
        ("some calibration", True),
    ],
)
def test_sartorius_measurement_is_message(value, expected):
    from sartoriusb import _is_message

    result = _is_message(value)

    assert result == expected


@pytest.mark.parametrize(
    "value,expected",
    [
        ("", ""),
        ("1", "1"),
        ("1]3", "1]3"),
        ("[", ""),
        ("1[2", "12"),
        ("1[2]", "12  "),
        ("1[2] x", "12   x"),
    ],
)
def test_sartorius_remove_calibration_note(value, expected):
    from sartoriusb import _remove_calibration_note

    result = _remove_calibration_note(value)

    assert result == expected


@pytest.mark.parametrize(
    "text,value,unit,stable,cal",
    [
        (" 123.45678 g  ", "123.45678", "g", True, True),
        (" 123.4567[8]g  ", "123.45678", "g", True, False),
        ("+123.45678 mg  ", "+123.45678", "mg", True, True),
        ("-  3.456 mg  ", "-3.456", "mg", True, True),
        ("-  3.456     ", "-3.456", None, False, True),
        (" 0.0   ", "0.0", None, False, True),
    ],
)
def test_sartorius_parse_16_char_output(text, value, unit, stable, cal):
    from sartoriusb import _parse_16_char_output

    result = _parse_16_char_output(text, mode="X")

    assert result == ("X", value, unit, stable, None)


def test_sartorius_parse_16_char_output_on_message(mocker):
    from sartoriusb import _parse_16_char_output

    result = _parse_16_char_output(" some Error ")

    assert result == (None, None, None, None, "some Error")


tests/test_pysartoriusb.py
import pytest

from unittest.mock import call


def test_sartorius_usb_init():
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb(1, 2, timeout=3, something=4, different=6)

    assert sub._serial_args == (1, 2)
    assert sub._serial_kargs == {"timeout": 3, "something": 4, "different": 6}
    assert sub._con is None


@pytest.mark.parametrize("con_value,expected", [(None, 1), ("Something", 0)])
def test_sartorius_usb_connection_property(mocker, con_value, expected):
    from pysartoriusb import SartoriusUsb

    mocker.patch.object(SartoriusUsb, "open")
    sub = SartoriusUsb()
    sub._con = con_value

    con = sub.connection

    assert sub.open.call_count == expected
    assert con == con_value


def test_sartorius_usb_parameters_passed_to_serial(mocker):
    from pysartoriusb import SartoriusUsb, serial

    mocker.patch.object(serial, "Serial", return_value="dummy connection")
    sub = SartoriusUsb(1, 2, something=4, different=6)

    sub.open()

    assert serial.Serial.call_count == 1
    assert serial.Serial.call_args == call(
        1, 2, timeout=1, something=4, different=6
    )
    assert sub._con == "dummy connection"


def test_sartorius_usb_close_calls_close_on_established_connection(mocker):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    sub._con = mocker.Mock()
    mocked_con = sub._con

    sub.close()

    assert sub._con is None
    assert mocked_con.close.call_count == 1


@pytest.mark.parametrize(
    "value,expected", [(b"bytes", b"bytes"), ("string", b"string")]
)
def test_sartorius_sends_bytes(mocker, value, expected):
    from pysartoriusb import SartoriusUsb, ESC, CR, LF

    sub = SartoriusUsb()
    sub._con = mocker.Mock()

    sub.send(value)

    assert sub._con.write.call_count == 1
    assert sub._con.write.call_args == call(ESC + expected + CR + LF)


def test_sartorius_read_passes_argument(mocker):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    sub._con = mocker.Mock()
    sub._con.read.return_value = "Hello!"

    result = sub.read(42)

    assert sub._con.read.call_count == 1
    assert sub._con.read.call_args == call(42)
    assert result == "Hello!"


def test_sartorius_readline(mocker):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    sub._con = mocker.Mock()
    sub._con.readline.return_value = "Hello!"

    result = sub.readline()

    assert sub._con.readline.call_count == 1
    assert sub._con.readline.call_args == call()
    assert result == "Hello!"


@pytest.mark.parametrize("stop_line", [b"", b" "])
@pytest.mark.parametrize(
    "stop_position,expected",
    [(0, []), (1, ["one"]), (2, ["one", "two"]), (3, ["one", "two", "three"])],
)
def test_sartorius_readlines(mocker, stop_line, stop_position, expected):
    from pysartoriusb import SartoriusUsb

    data = [b"one", b"two", b"three", b"four", b"five"]
    data[stop_position] = stop_line

    sub = SartoriusUsb()
    sub._con = mocker.Mock()
    sub._con.readline.side_effect = data

    result = sub.readlines()

    assert sub._con.readline.call_count == stop_position + 1
    assert result == expected


def test_sartorius_get(mocker):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    mocker.patch.object(sub, "send")
    mocker.patch.object(sub, "readlines", return_value="some result")

    result = sub.get("some command")

    assert sub.send.call_count == 1
    assert sub.send.call_args == call("some command")
    assert sub.readlines.call_count == 1
    assert sub.readlines.call_args == call()
    assert result == "some result"


def test_sartorius_measure_with_data(mocker):
    from pysartoriusb import SartoriusUsb, CMD_PRINT

    sub = SartoriusUsb()
    mocker.patch.object(sub, "get", return_value=["value", "some other stuff"])
    mocker.patch.object(sub, "parse_measurement", return_value="parsed result")

    result = sub.measure()

    assert sub.get.call_count == 1
    assert sub.get.call_args == call(CMD_PRINT)
    assert sub.parse_measurement.call_count == 1
    assert sub.parse_measurement.call_args == call("value")
    assert result == "parsed result"


def test_sartorius_measure_with_timeout(mocker):
    from pysartoriusb import SartoriusUsb, CMD_PRINT

    sub = SartoriusUsb()
    mocker.patch.object(sub, "get", return_value=[])
    mocker.patch.object(sub, "parse_measurement", return_value="parsed result")

    result = sub.measure()

    assert sub.get.call_count == 1
    assert sub.get.call_args == call(CMD_PRINT)
    assert sub.parse_measurement.call_count == 0
    assert result == (
                None, None, None, None, "Connection Timeout", None
            )

@pytest.mark.parametrize(
    "value,count_16,count_22,expected",
    [
        ("123456789012345", 1, 0, "16"),
        ("1234567890123456", 1, 0, "16"),
        ("12345678901234567", 0, 1, "22"),
    ],
)
def test_sartorius_parse_measurement(
    mocker, value, count_16, count_22, expected
):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    mocker.patch.object(sub, "_parse_16_char_output", return_value="16")
    mocker.patch.object(sub, "_parse_22_char_output", return_value="22")

    result = sub.parse_measurement(value)

    assert sub._parse_16_char_output.call_count == count_16
    assert sub._parse_22_char_output.call_count == count_22
    assert result == expected


def test_sartorius_parse_22_char_output(mocker):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    mocker.patch.object(sub, "_parse_16_char_output", return_value="16")

    result = sub._parse_22_char_output(" mode value ")

    assert sub._parse_16_char_output.call_count == 1
    assert sub._parse_16_char_output.call_args == call("value ", "mode")
    assert result == "16"


@pytest.mark.parametrize(
    "value,expected",
    [
        ("12345", False),
        (" some text ", False),
        (" Error 123 ", "Error 123"),
        ("highlander ", "highlander"),
        (" too low", "too low"),
        ("123--456", "123--456"),
        ("some calibration", "some calibration"),
    ],
)
def test_sartorius_measurement_is_message(value, expected):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    result = sub._is_message(value)

    assert result == expected


@pytest.mark.parametrize(
    "value,data,cal",
    [
        ("", "", True),
        ("1", "1", True),
        ("1]3", "1]3", True),
        ("[", "", False),
        ("1[2", "12", False),
        ("1[2]", "12  ", False),
        ("1[2] x", "12   x", False),
    ],
)
def test_sartorius_adjust_raw_data_for_calibration_note(value, data, cal):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    result = sub._adjust_raw_data_for_calibration_note(value)

    assert result == (data, cal)


@pytest.mark.parametrize(
    "text,value,unit,stable,cal",
    [
        (" 123.45678 g  ", "123.45678", "g", True, True),
        (" 123.4567[8]g  ", "123.45678", "g", True, False),
        ("+123.45678 mg  ", "+123.45678", "mg", True, True),
        ("-  3.456 mg  ", "-3.456", "mg", True, True),
        ("-  3.456     ", "-3.456", None, False, True),
        (" 0.0   ", "0.0", None, False, True),
    ],
)
def test_sartorius_parse_16_char_output(text, value, unit, stable, cal):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    result = sub._parse_16_char_output(text, mode="X")

    assert result == ("X", value, unit, stable, False, cal)


def test_sartorius_parse_16_char_output_on_message(mocker):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    mocker.patch.object(sub, "_adjust_raw_data_for_calibration_note")
    result = sub._parse_16_char_output(" some Error ")

    assert result == (None, None, None, None, "some Error", None)


def test_sartorius_context_manager(mocker):
    from pysartoriusb import SartoriusUsb

    sub = SartoriusUsb()
    mocker.patch.object(sub, "open")
    mocker.patch.object(sub, "close")

    with sub as context:
        assert sub is context
        assert sub.open.call_count == 1
        assert sub.close.call_count == 0

    assert sub.open.call_count == 1
    assert sub.close.call_count == 1


sartoriusb/__init__.py
""" pysartoriusb

Controlling a Sartorius Quintix Scale via USB

"""

__version__ = "0.2.5"

from collections import namedtuple

import serial

CMD_PRINT = b"P"
CMD_TARA = b"T"

CMD_EXPCLICIT_TARA = b"U"
CMD_EXPCLICIT_NULL = b"V"

CMD_INFO_TYPE = b"x1_"
CMD_INFO_SNR = b"x2_"
CMD_INFO_VERSION_SCALE = b"x3_"
CMD_INFO_VERSION_CONTROL_UNIT = b"x4_"
CMD_INFO_USER = b"x5_"

CMD_FILTER_ENVIRONMENT_VERY_STABLE = b"K"
CMD_FILTER_ENVIRONMENT_STABLE = b"L"
CMD_FILTER_ENVIRONMENT_UNSTABLE = b"M"
CMD_FILTER_ENVIRONMENT_VERAY_UNSTABLE = b"N"

CMD_KEYBOARD_LOCK = b"O"
CMD_KEYBOARD_UNKLOCK = b"R"
CMD_KEYPRESS_PRINT = b"kP_"
CMD_KEYPRESS_CANCEL = b"s3_"

CMD_BEEP = b"Q"

CMD_RESTART = b"S"
CMD_ADJUST_INTERNAL = b"Z"
CMD_ADJUST_EXTERNAL = b"W"


ESC = chr(27).encode("utf-8")
CR = chr(13).encode("utf-8")
LF = chr(10).encode("utf-8")


Measurement = namedtuple(
    "Measurement", ["mode", "value", "unit", "stable", "message"],
)


class SartoriusUsb:

    encoding = "ascii"

    def __init__(self, *serial_args, timeout=1, **serial_kargs):
        """ initialization fo the class

        All arguments are passed on to serial.Serial

        :params *serial_args: positional arguments for serial.Serial
        :params timeout: timeout of connection in seconds, defaults to 1
        :params *serial_kargs: keyword arguments for serial.Serial
        """

        self._serial_args = serial_args

        self._serial_kargs = serial_kargs
        self._serial_kargs["timeout"] = timeout

        self._con = None

    @property
    def connection(self):
        """ returns the serial connection, establishes a new one if needed """
        if self._con is None:
            self.connect()
        return self._con

    def connect(self):
        """ establishes a new serial connection """
        if self._con is None:
            self._con = serial.Serial(*self._serial_args, **self._serial_kargs)

    def open(self):
        """ establishes a new serial connection

        This function just calls the 'connect()' method and is here for
        compability with other libraries that use open() / close()
        """
        self.connect()

    def close(self):
        """ closes a serial connection, if one is open """
        if self._con:
            self._con.close()
            self._con = None

    def send(self, command):
        """ sends a command to the scale """
        if not isinstance(command, bytes):
            command = str(command).encode(self.encoding)
        self.connection.write(ESC + command + CR + LF)

    def read(self, nr_of_bytes=1):
        """ reads some number of bytes from the serial connection """
        return self.connection.read(nr_of_bytes)

    def readline(self):
        """ reads bytes from the serial connection until a newline """
        return self.connection.readline()

    def readlines(self):
        """ returns a list of lines of available data """
        lines = []
        while True:
            line = self.connection.readline()
            line = line.decode(self.encoding)
            if not line.strip():
                # a line with only whitespace shows the end of the output
                # also a read timeout produces b''
                break
            lines.append(line)
        return lines

    def get(self, command):
        """ sends a command and returns the available data """
        self.send(command)
        return self.readlines()

    def measure(self):
        """ sends a print"""
        raw_data_lines = self.get(CMD_PRINT)
        if raw_data_lines:
            raw_data = raw_data_lines[0]
            return parse_measurement(raw_data)
        else:
            # propably serial connection timeout
            return Measurement(None, None, None, None, "Connection Timeout")

    def __enter__(self):
        """ Context manager: establishes connection """
        self.connect()
        return self

    def __exit__(self, exc_type, exc_value, exc_traceback):
        """ Context manager: closes connection """
        self.close()


def parse_measurement(raw_data):
    """ parses the raw data from a measurement """
    if len(raw_data) <= 16:
        return _parse_16_char_output(raw_data)
    else:
        return _parse_22_char_output(raw_data)


def _parse_22_char_output(raw_data):
    """ parse a 16 character measurement output

    The scale can be set to return two different types of output. This
    function parses a 16 character output.
    """

    mode = raw_data[:6].strip()
    rest = raw_data[6:]

    return _parse_16_char_output(rest, mode)


def _parse_16_char_output(raw_data, mode="unknown"):
    """ parse a 16 character measurement output

    The scale can be set to return two different types of output. This
    function parses a 16 character output.
    """

    if _is_message(raw_data):
        msg = raw_data.strip()
        return Measurement(None, None, None, None, msg)

    raw_data = _remove_calibration_note(raw_data)

    sign = raw_data[0].strip()
    value_and_unit = raw_data[1:].strip()
    parts = value_and_unit.rsplit(" ", 1)
    raw_value = parts[0]
    value = sign + "".join(raw_value.split())

    if len(parts) == 2:
        unit = parts[1]
        stable = True
    else:
        unit = None
        stable = False

    return Measurement(mode, value, unit, stable, None)


def _is_message(raw_data):
    """ returns the message that occured in a measurement or False """
    for identifier in ("high", "low", "cal", "err", "--"):
        if identifier in raw_data.lower():
            return True
    return False


def _remove_calibration_note(raw_data):
    """ adjusts the raw data string if a calibration note is present

    According to the manual, this should not happen in SBI mode of the
    scale. This is included to prevent hiccups but probably not handled
    the right way....

    The data with a calibration node on in put and output of this method

    in:  "+123.4567[8]g  "
    out: "+123.45678  g  "

    """
    if "[" in raw_data:
        raw_data = raw_data.replace("[", "").replace("]", "  ")
    return raw_data
